### DISCUSSION TITLE
Postgres Language Server: JS + TS Template Literal Support

### DISCUSSION THREAD

**Discussion by karlhorky** [supabase-community/postgres-language-server#25] [Ideas]
Originally opened as issue and a discussion in the organization repo (probably the wrong locations):

- Issue https://github.com/supabase/postgres_lsp/issues/20
- Organization-level discussion https://github.com/orgs/supabase/discussions/16286

---

Hi there 👋 First of all, thanks for all of the effort that Supabase is putting into the PostgreSQL community, including tooling, really amazing 🙌 

I recently encountered Postgres Language Server via [the recent tweet](https://twitter.com/karlhorky/status/1688489997240881152) from @kiwicopple and it looks amazing!

I have been chasing the "holy grail" of tooling for SQL template literals in JavaScript + TypeScript for a while now:

- syntax highlighting (works, with the VS Code extension [SQL tagged template literals (syntax only)](https://marketplace.visualstudio.com/items?itemName=frigus02.vscode-sql-tagged-template-literals-syntax-only))
- linting (works, using [SafeQL](https://github.com/ts-safeql/safeql/))
- formatting, some options
  - [`prettier-plugin-embed` and `prettier-plugin-sql`](https://github.com/Sec-ant/prettier-plugin-embed/issues/20) - uses [`sql-formatter`](https://github.com/sql-formatter-org/sql-formatter)
  - [kind of works, with `eslint-plugin-sql` and `eslint-plugin-unicorn`](https://github.com/gajus/eslint-plugin-sql/pull/33#issuecomment-1666585459), but very opinionated / not so configurable - uses [`pg-formatter`](https://github.com/gajus/pg-formatter)
  - [`@potygen/prettier-plugin-pgsql`](https://github.com/ivank/potygen/blob/main/packages/prettier-plugin-pgsql/README.md) by @ivank - uses custom Prettier AST formatter

Wasn't even looking for inline documentation yet, but that's a great addition too!

It seems that the plan will be for Postgres Language Server to support all of these use cases, which is amazing!

I would love to have everything handled by one highly-battle-tested tool, which could maybe also deal with some of the shortcomings from the tools listed above.

However, for Postgres Language Server to work for any of our projects, it would need a feature which I'm not sure it currently has (can't tell, from a quick glance at the docs and the code):

- Support for tagged template literals in JavaScript and TypeScript

There is precedent for SQL tagged template literals in JS+TS, some of the more popular ones being:

- [Postgres.js](https://github.com/porsager/postgres)
- [`@vercel/postgres`](https://vercel.com/docs/storage/vercel-postgres/sdk)

Eg. I would love for all of the features of Postgres Language Server to be available in the following code block:

```ts
import { createClient } from '@vercel/postgres';
 
async function queryPosts() {
  const client = createClient();
  await client.connect();
 
  try {
    const likes = 100;
    const { rows, fields } =
      await client.sql`SELECT * FROM posts WHERE likes > ${likes};`;
  } finally {
    await client.end();
  }
}
```

Or this code block:

```ts
// db.js
import postgres from 'postgres'

const sql = postgres({ /* options */ }) // will use psql environment variables

async function getUsersOver(age) {
  const users = await sql`
    select
      name,
      age
    from users
    where age > ${ age }
  `
  // users = Result [{ name: "Walter", age: 80 }, { name: 'Murray', age: 68 }, ...]
  return users
}


async function insertUser({ name, age }) {
  const users = await sql`
    insert into users
      (name, age)
    values
      (${ name }, ${ age })
    returning name, age
  `
  // users = Result [{ name: "Murray", age: 68 }]
  return users
}
```

**Comment by psteinroe** [comment]
hey @karlhorky, sorry for the late reply!

I have been doing some research on this in the meanwhile and it seems like that it's the client who has to support multiple language servers for a single file. You can find a related discussion [here](https://news.ycombinator.com/item?id=37023959). for vscode as a client it is described [here](https://code.visualstudio.com/api/language-extensions/embedded-languages). I also found a [neovim plugi ](https://github.com/jmbuhr/otter.nvim) that implements something similar. In summary, it will require some changes to the typescript plugin / extension. I am quite surprised myself that almost all editors (exception being Jetbrains products) only support a single language server per file.

**Comment by remorses** [comment]
This would be amazing

It's possible to add completions for SQL inside Typescript and Js files with a typescript plugin, this is an example: https://github.com/runem/lit-analyzer 

(notice that there is another similar extension but autocompletion is not working there)

**Reply by karlhorky** [reply]
Nice find! I guess this works via `ts-lit-plugin` created by @runem and co-maintained by @rictic:

> Typescript plugin that adds type checking and code completion to lit-html

- https://github.com/runem/lit-analyzer/tree/master/packages/ts-lit-plugin

**Reply by karlhorky** [reply]
Oh I also found one that does completions inside SQL inside of tagged template literals in JS/TS, by @xialvjun and @darky:

- https://github.com/xialvjun/ts-sql-plugin

**Reply by remorses** [reply]
No `lit-analyzer` is different from `ts-lit-plugin`, as i said completions in `ts-lit-plugin` don't work

Also `lit-analyzer` has the same interface as a typical LSP so it should be easy to use as starting point

**Reply by karlhorky** [reply]
Ah do you mean [`completionsAtOffset()`](https://github.com/runem/lit-analyzer/blob/a8278e2d71dc508bcdc5d218afe54ae558a8864c/packages/lit-analyzer/src/lib/analyze/document-analyzer/html/completion/completions-at-offset.ts#L11-L40) in `packages/lit-analyzer/src/lib/analyze/document-analyzer/html/completion/completions-at-offset.ts`? I'm not as familiar as I would like to be with LSPs...

**Comment by hatemhosny** [comment]
I'm not sure if that helps, but [Volar.js](https://volarjs.dev/) simplifies handling [embedded languages](https://volarjs.dev/core-concepts/embedded-languages/).
